platform/core.multitabs/src/org/netbeans/core/multitabs/impl/DocumentSwitcherTable.java

```diff
@@ -152,7 +152,21 @@ boolean onMouseEvent( MouseEvent e ) {
                         TabActionEvent tae = new TabActionEvent( this, TabbedContainer.COMMAND_CLOSE, tabIndex );
                         controller.postActionEvent( tae );
                         return true;
+                    }else if(tabIndex == -1){
+                        List<TabData> tabs = controller.getTabModel().getTabs();
+                        ProjectProxy project = item.project;
+                        ProjectSupport projectSupport = ProjectSupport.getDefault();
+                        for(int i = tabs.size(); i-- > 0;){
+                            TabData projectTab = tabs.get(i);
```

Comment:
Yeah, when closing tab one by one, the tabIndex changes which resulted in closing only the first tab under document list. Managed to solve that by closing tab from the end

Reply:
Makes sense. Would be good to leave a comment there so that nobody changes it by accident.

Reply:
> Please try to extract the new code which closes the tabs into a package private utility method and put it into DocumentSwitcherTable . Since ButtonPupupSwitcher is in the same package as DocumentSwitcherTable and can simply call its method without code duplication.

Somehow I pushed older version of the code, difference should be visible in the next commit. When closing tabs using DELETE key, I have to call SwingUtilities if all the tabs are closed. Otherwise the popup stays open. That is why I am keeping it outside of shared method.

Reply:
if you would move this after the loop, it should still work and we could extract the whole section into a method in `DocumentSwitcherTable`? Duplicated code is really something we should try to avoid.

Reply:
Oh yeah, you are right.

Not sure if I have done it correctly, is this the usage in ButtonPopupSwitcher?
```
DocumentSwitcherTable dst = new DocumentSwitcherTable( controller, createSwitcherItems(controller), y );
dst.closeDocumentList( item, controller );
if( controller.getTabModel().size() == 1 ) {
    SwingUtilities.invokeLater(this::hideCurrentPopup);
}
```

Reply:
i commented inline below. closing this thread.